src/approaches/pg3_approach.py

```diff
@@ -0,0 +1,429 @@
+                                      for task in train_tasks]
+
+    def __call__(self, ldl: LiftedDecisionList) -> float:
+        """Compute the heuristic value for the given LDL policy."""
+        score = 0.0
+        for idx in range(len(self._abstract_train_tasks)):
+            score += self._get_score_for_task(ldl, idx)
+        logging.debug(f"Scoring:\n{ldl}\nScore: {score}")
+        return score
+
+    @abc.abstractmethod
+    def _get_score_for_task(self, ldl: LiftedDecisionList,
+                            task_idx: int) -> float:
+        """Produce a score, with lower better."""
+        raise NotImplementedError("Override me!")
+
+
+class _PolicyEvaluationPG3Heuristic(_PG3Heuristic):
+    """Score a policy based on the number of train tasks it solves at the
+    abstract level."""
+
+    def _get_score_for_task(self, ldl: LiftedDecisionList,
+                            task_idx: int) -> float:
+        _, atoms, goal = self._abstract_train_tasks[task_idx]
+        if self._ldl_solves_abstract_task(ldl, atoms, goal):
+            return 0.0
+        return 1.0
+
+    @staticmethod
+    def _ldl_solves_abstract_task(ldl: LiftedDecisionList,
+                                  atoms: Set[GroundAtom],
+                                  goal: Set[GroundAtom]) -> bool:
+        for _ in range(CFG.horizon):
+            if goal.issubset(atoms):
+                return True
+            ground_nsrt = utils.query_ldl(ldl, atoms, goal)
+            if ground_nsrt is None:
+                return False
+            atoms = utils.apply_operator(ground_nsrt, atoms)
+        return goal.issubset(atoms)
+
+
+class _PlanComparisonPG3Heuristic(_PG3Heuristic):
+    """Score a policy based on agreement with certain plans."""
+
+    def _get_score_for_task(self, ldl: LiftedDecisionList,
+                            task_idx: int) -> float:
+        atom_plan = self._get_atom_plan_for_task(ldl, task_idx)
+        # Note: we need the goal because it's an input to the LDL policy.
+        _, _, goal = self._abstract_train_tasks[task_idx]
+        assert goal.issubset(atom_plan[-1])
+        return self._count_missed_steps(ldl, atom_plan, goal)
+
+    @abc.abstractmethod
+    def _get_atom_plan_for_task(self, ldl: LiftedDecisionList,
+                                task_idx: int) -> Sequence[Set[GroundAtom]]:
+        """Given a task, get the plan with which we will compare the policy."""
+        raise NotImplementedError("Override me!")
+
+    @staticmethod
+    def _count_missed_steps(ldl: LiftedDecisionList,
+                            atom_seq: Sequence[Set[GroundAtom]],
+                            goal: Set[GroundAtom]) -> float:
+        missed_steps = 0.0
+        for t in range(len(atom_seq) - 1):
+            ground_nsrt = utils.query_ldl(ldl, atom_seq[t], goal)
+            if ground_nsrt is None:
+                missed_steps += CFG.pg3_plan_compare_inapplicable_cost
```

Comment:
yeah this is something we didn't do in the original implementation and I was playing around with here. a lower value seems to help because it means that if policy A is "confidently incorrect" and policy B is "not sure", we prefer policy B. this is maybe good because the policy search goes from general to specific, so there is kinda more hope for policy B than A in the future of search. but we will have to experiment at scale to see whether this actually helps.